src/bin/cratesfyi.rs

```diff
@@ -476,5 +490,51 @@ impl BlacklistSubcommand {
             Self::Remove { crate_name } => db::blacklist::remove_crate(&conn, &crate_name)
                 .expect("failed to remove crate from blacklist"),
         }
+        Ok(())
+    }
+}
+
+struct Context {
+    build_queue: OnceCell<Arc<BuildQueue>>,
+    config: OnceCell<Arc<Config>>,
+    pool: OnceCell<Pool>,
+}
+
+impl Context {
+    fn new() -> Self {
+        Self {
+            build_queue: OnceCell::new(),
+            config: OnceCell::new(),
+            pool: OnceCell::new(),
+        }
+    }
+
+    fn build_queue(&self) -> Result<Arc<BuildQueue>, Error> {
+        Ok(self
+            .build_queue
+            .get_or_try_init::<_, Error>(|| {
+                Ok(Arc::new(BuildQueue::new(self.pool()?, &*self.config()?)))
+            })?
+            .clone())
+    }
+
+    fn config(&self) -> Result<Arc<Config>, Error> {
+        Ok(self
+            .config
+            .get_or_try_init::<_, Error>(|| Ok(Arc::new(Config::from_env()?)))?
+            .clone())
+    }
+
+    fn pool(&self) -> Result<Pool, Error> {
+        Ok(self
+            .pool
+            .get_or_try_init::<_, Error>(|| Ok(Pool::new(&*self.config()?)?))?
+            .clone())
+    }
```

Comment:
This function is called a lot. Would it be better to remove the once_cell in favor of constructing the pool eagerly? That way it doesn't have to wait for a mutex lock each time.

I guess it only needs to lock the mutex when initializing for the first time, so it wouldn't be a problem in prod, but it would make tests even more painful than currently since they all run in parallel.

Reply:
`once_cell` is not a Mutex. Also, I don't see how this would be a problem with tests: tests don't use this struct, and they don't share any state or Mutex except for metrics and the handlebars rustc version.

Reply:
https://docs.rs/crate/once_cell/1.4.0/source/src/imp_pl.rs

It uses a mutex when initializing. So if many threads try to initialize at once they might fight over the lock.

Since this doesn't run in tests though it shouldn't be an issue as we call `pool()` pretty early.